plugins/outputs/nats/nats_test.go

```diff
@@ -263,3 +266,230 @@ func createStream(t *testing.T, server string, cfg *nats.StreamConfig) {
+				WaitingFor:   wait.ForListeningPort(nat.Port(natsServicePort)),
+			},
+			nats: &NATS{
+				Name:    "telegraf",
+				Subject: "my-subject.{{ .Name }}.{{ .Tag \"tag1\" }}.{{ .Tag \"tag2\" }}",
+				Jetstream: &StreamConfig{
+					Name: "my-telegraf-stream",
+				},
+				serializer: &influx.Serializer{},
+				Log:        testutil.Logger{},
+			},
+			wantErr: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			err := tc.container.Start()
+			require.NoError(t, err, "failed to start container")
+			defer tc.container.Terminate()
+
+			server := []string{fmt.Sprintf("nats://%s:%s", tc.container.Address, tc.container.Ports[natsServicePort])}
+			tc.nats.Servers = server
+			// Verify that we can connect to the NATS daemon
+			err = tc.nats.Init()
+			if tc.wantErr {
+				require.Error(t, err)
+				return
+			}
+			err = tc.nats.Connect()
+			require.NoError(t, err)
+
+			if tc.nats.Jetstream != nil {
+				stream, err := tc.nats.jetstreamClient.Stream(t.Context(), tc.nats.Jetstream.Name)
+				require.NoError(t, err)
+				si, err := stream.Info(t.Context())
+				require.NoError(t, err)
+
+				tc.streamConfigCompareFunc(t, si)
+			}
+			// Verify that we can successfully write data to the NATS daemon
+			metric := testutil.MockMetrics()
+			for _, m := range metric {
+				for k, v := range tc.tags {
+					m.AddTag(k, v)
+				}
+				for k, v := range tc.fields {
+					m.AddField(k, v)
+				}
+			}
+
+			err = tc.nats.Write(metric)
+			require.NoError(t, err)
+
+			foundSubjects := make([]string, 0)
+			if tc.nats.Jetstream != nil {
+				js, err := tc.nats.conn.JetStream()
+				require.NoError(t, err)
+				sub, err := js.PullSubscribe(tc.nats.Jetstream.Subjects[0], "")
+				require.NoError(t, err)
+
+				msgs, err := sub.Fetch(100, nats.MaxWait(1*time.Second))
+				if err != nil {
+					if !errors.Is(err, nats.ErrTimeout) {
+						require.NoError(t, err)
+					}
+				}
+
+				require.Len(t, msgs, tc.msgCount, "unexpected number of messages")
+				for _, msg := range msgs {
+					foundSubjects = append(foundSubjects, msg.Subject)
+				}
+			}
+
+			err = validateSubjects(tc.expectedSubjects, foundSubjects)
+			require.NoError(t, err)
+		}) // end of test case
+	}
+}
+
+// validateSubjects checks that:
+// - All entries in generated exist in expected.
+// - All expected values appear at least once in generated.
+// Returns error if either condition fails.
+func validateSubjects(expected, generated []string) error {
+	expectedSet := make(map[string]struct{}, len(expected))
+	for _, e := range expected {
+		expectedSet[e] = struct{}{}
+	}
+
+	seen := make(map[string]bool)
+
+	for _, g := range generated {
+		if _, ok := expectedSet[g]; !ok {
+			return fmt.Errorf("invalid entry found: %q is not in expected list", g)
+		}
+		seen[g] = true
+	}
+
+	for _, e := range expected {
+		if !seen[e] {
+			return fmt.Errorf("missing expected entry: %q not found in generated list", e)
+		}
+	}
+
+	return nil
+}
```

Comment:
Why not use [`require.Equal`](https://pkg.go.dev/github.com/stretchr/testify/require#Equal) or, if the order may vary, [`require.ElementsMatch`](https://pkg.go.dev/github.com/stretchr/testify/require#ElementsMatch)?